xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            - Cannot be set if `pat` is a compiled regex
+
+        Returns
+        -------
+        extracted : same type as values or object array
+
+        Raises
+        ------
+        ValueError : `pat` has no capture groups.
+        ValueError : `case` is set when `pat` is a compiled regular expression.
+        KeyError : Either of the given dimensions is already present in the DataArray.
+        KeyError : The given dimensions names are the same.
+
+        Examples
+        --------
+        Create a string array
+
+        >>> value = xr.DataArray(
+        ...     [
+        ...         [
+        ...             "a_Xy_0",
+        ...             "ab_xY_10-bab_Xy_110-baab_Xy_1100",
+        ...             "abc_Xy_01-cbc_Xy_2210",
+        ...         ],
+        ...         [
+        ...             "abcd_Xy_-dcd_Xy_33210-dccd_Xy_332210",
+        ...             "",
+        ...             "abcdef_Xy_101-fef_Xy_5543210",
+        ...         ],
+        ...     ],
+        ...     dims=["X", "Y"],
+        ... )
+
+        Extract matches
+
+        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="XX", match_dim="YY")
+        <xarray.DataArray (X: 2, Y: 3, XX: 3, YY: 2)>
+        array([[[['a', '0'],
+                 ['', ''],
+                 ['', '']],
+        <BLANKLINE>
+                [['bab', '110'],
+                 ['baab', '1100'],
+                 ['', '']],
+        <BLANKLINE>
+                [['abc', '01'],
+                 ['cbc', '2210'],
+                 ['', '']]],
+        <BLANKLINE>
+        <BLANKLINE>
+               [[['abcd', ''],
+                 ['dcd', '33210'],
+                 ['dccd', '332210']],
+        <BLANKLINE>
+                [['', ''],
+                 ['', ''],
+                 ['', '']],
+        <BLANKLINE>
+                [['abcdef', '101'],
+                 ['fef', '5543210'],
+                 ['', '']]]], dtype='<U7')
+        Dimensions without coordinates: X, Y, XX, YY
+
+
+        See Also
+        --------
+        DataArray.str.extract
+        DataArray.str.findall
+        re.compile
+        re.findall
+        pandas.Series.str.extractall
+        """
+        pat = self._re_compile(pat, flags, case)
+
+        if pat.groups == 0:
+            raise ValueError("No capture groups found in pattern.")
+
+        if group_dim in self._obj.dims:
+            raise KeyError(f"Group dimension {group_dim} already present in DataArray.")
+
+        if match_dim in self._obj.dims:
+            raise KeyError(f"Match dimension {match_dim} already present in DataArray.")
+
+        if group_dim == match_dim:
+            raise KeyError(
+                f"Group dimension {group_dim} is the same as match dimension {match_dim}."
+            )
+
+        _get_count = lambda x: len(pat.findall(x))
+        maxcount = self._apply(_get_count, dtype=np.int_).max().data.tolist()
+
+        def _get_res(val, pat=pat, num=maxcount, dtype=self._obj.dtype):
```

Comment:
```suggestion
        def _get_res(val):
```
I think you can remove the args - `_get_res` is in the namespace of `extractall`

Reply:
This guarantees that the names are bound with the correct values and not with other values that may be set later.  It prevents potential gotchas if variables are modified.